### DISCUSSION TITLE
Interesting applications

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#78] [General]
Starting this thread so we can collect interesting examples which can inspire or guide us. Please contribute your suggestions.

I'll start with this video from Sorin: https://www.youtube.com/watch?v=9HI8FerKr6Q

**Comment by ghost** [comment]


**Reply by kaveh808** [reply]
Yes, shape sweeping and I assume boolean operations on meshes, which can be complex. If someone is motivated to do this, I'm happy to provide guidance.

What parts of the workbench idea do you find interesting? I have been thinking about ways of organizing plugin functionality.

**Reply by kaveh808** [reply]
Interesting. I have been thinking along those lines, with some of the plugin files installing specialized menus into the GUI. e.g. `uv-mesh.lisp`

**Comment by awolven** [comment]
I've written cl bindings for opencascade which is the solid modeler that freecad uses. ***@***.***:awolven/oc.gitSent from my Galaxy
-------- Original message --------From: Marengo Zed ***@***.***> Date: 11/13/22  4:16 PM  (GMT-06:00) To: kaveh808/kons-9 ***@***.***> Cc: Subscribed ***@***.***> Subject: Re: [kaveh808/kons-9] Interesting applications (Discussion #78) 

My most recent thinking

It's been over a year since I looked at this stuff, so not very recent, and I'm obviously no expert either.  I think it's worth some of my time looking at exactly how FreeCAD and CAMotics have done it--I can't remember what techniques they use.  If there is anyone knowledgeable out there, please feel free to chime in.

**Comment by WillAdams** [comment]
I've put a tiny bit of code up on this for OpenSCAD/RapCAD, focused on making a 3D model based on a toolpath and then writing out G-code to actually cut the model:

https://github.com/WillAdams/gcodepreview

One thing of note is the issue of tool specification --- CAMotics uses an XML file for this, but there's a de facto industry standard of putting the tool description in comments before the tool change --- the PDF documentation for CutViewer writes this up quite nicely, and it allows for description of most tools.

One notable capability to consider is whether or no undercuts are allowed/modeled properly --- doing so allows one to represent keyhole cutters, dovetail cutters, and "lollipop" cutters which would be quite nice.

**Comment by awolven** [comment]
I have a generative knowledge based engineering kernel that I'm going to
augment kons-9 with, but that kernel is GPL https://github.com/awolven/adhoc
It is designed to be like Icad.

**Comment by awolven** [comment]
What's implicit function modeling?

**Reply by kaveh808** [reply]
Check out `demo-isosurface.lisp`.

**Reply by kaveh808** [reply]
Let me know if you have any questions. There isn't a proper plugin interface yet -- you just load your code and make kons-9 calls.

**Comment by awolven** [comment]
Adhoc itself does not contain any geometry.

**Comment by awolven** [comment]
Lisp was the original "plugin" language.

**Comment by ghost** [comment]


**Reply by kaveh808** [reply]
Neat seeing an app built on Symbolics machines. I'll peruse this over the weekend. Thanks.

**Reply by WillAdams** [reply]
As a person who has raged against it not being possible to position a point at exact coordinates (the software insisted on something like 10.0001 pts.), let me say, "Yes, please."

**Reply by foretspaisibles** [reply]
Exact arithmetic is occasionally suitable for geometric problems but if we are interested in drawing implicit curves or surfaces, doing riemannian geometry (*e.g.* distances on surfaces), animations based on physical laws (exponential), then exact arithmetic is just unable to do that. (Unless maybe everything is a [continued fraction](https://en.wikipedia.org/wiki/Continued_fraction).)

**Reply by aykaramba** [reply]
Wouldn't it be cool if we could get Greenspun and some Symbolics people on the project just for fun?

**Comment by awolven** [comment]
Build it on Bill Gosper's continued fraction expansions.

**Comment by shalindesigns2023** [comment]


**Reply by WillAdams** [reply]
Apparently SPAM bots have made their way to Github. Could @kaveh808 or someone tidy that up? Did report it from their user page.

**Reply by kaveh808** [reply]
Deleted.

**Comment by ghost** [comment]


**Reply by kaveh808** [reply]
There is very rudimentary support for user-defined interactions. See `test/demo-interactor.lisp` for some example snake games.

**Comment by foretspaisibles** [comment]
**Furnitures arrangement planner**

Add primitives to create rooms and furnitures, to help people model their flat and plan the arrangement of furnitures or changes. Could be handy before considerable changes(refitting a closet, expanding the kitchen, switching bed or sofa) or in tight places that.

**Comment by foretspaisibles** [comment]
**Physics engine**

Couple with a physics engine to generate realistic animations or models.

1. Bouncy cubes falling from a height on the ground.
2. Same as 1. with knobs to turn gravity on and off :-)
3. Animated solar system or planet with satellites, with Newton mechanics.
4. Air flowing through a fan and flowing a curtain.
5. Wind carrying autumn leaves.
6. Cars, buses, and trains interacting in a city (or toy city)
7. Electrical magnet.
8. Robot mouse finding the goal in a labyrinth (the robot has sensors and the engine let the robot bump into walls.)
9.  Materials whose resistance collapse.
10. Special frequencies of a bridge.

**Reply by kaveh808** [reply]
This may be a place where we can interface to an existing C/C++ physics engine. As always though, I am wary of adding dependencies which could complicate setup.

**Comment by foretspaisibles** [comment]
**Data Exploration**

Set great plotting and tabulating facilities, with export to common formats (raster PNG or SVG). Make stunning visualisations, and data driven documents. There is a heap of ideas under D3 (https://d3js.org)

**Comment by foretspaisibles** [comment]
**Common Lisp IDE**

Pull an Emacs-like editor to make a source code editor with incredible code navigation, and visualisation abilities (coverage, performance or value inspection). Take advantage of human spatial abilities.

**Reply by kaveh808** [reply]
I like the idea of visualization tools for inspecting code and data structures. Maybe starting with a UI on top of the `inspect` facility.

**Comment by ghost** [comment]


**Reply by kaveh808** [reply]
I remember that. :)

**Comment by ghost** [comment]


**Reply by kaveh808** [reply]
Very cool. On a side note, if anyone wants to add audio capabilities to kons-9, that would be neat. Unfortunately I know nothing about audio software or music.

But I can imagine algorithmically-generated music written in CL...

**Reply by ghost** [reply]
I would say clone me, I could do audio in my sleep, but one of me is enough.

**Reply by theottm** [reply]
There is [incudine](https://incudine.sourceforge.net).
[Someone](https://github.com/azimut/incudine-cepl) used CEPL with it for live coding. I never tried it but it looks promising.

I will try it out at some point when I go back to music production. If I have fun with I could have a try adding it to kons-9

**Reply by kaveh808** [reply]
I wonder if adding audio will require somehow ensuring "real time" playback. Currently we don't have such a thing. Animation is done by simply holding down the spacebar. Convenient but a kludge.

**Reply by theottm** [reply]
That's probably going to be a whole topic :) I never played with building a sound engine myself. Nore have I written something real time.

Two systems that I used in the past to make sound are exempore and SuperCollider. They both have a DSP server running in real time. 

Extempore is actually very interesting, it is written in scheme with a statically typed DSL for low level DSP code. It is designed for live coding. In some way it is the CEPL of music, since CEPL integrates gpu code in the main code as well and targets live coding as well.

Real-time or the management of time constraints is actually an relevant consideration even out of the scope of music, if someone wants to build a game on kons-9 for example.

**Reply by kaveh808** [reply]
We shall see in what direction(s) the system develops in organically over time. Discovery is part of the joy of the journey.

**Reply by theottm** [reply]
I'm curious to see where it leads

**Comment by ghost** [comment]


**Reply by kaveh808** [reply]
What part of this would be applicable to kons-9?

**Reply by kayomarz** [reply]
Nikola Tesla too used visualization. He used it to help him invent things .  His
power of visualization started with a childhood problem wherein images used to
appear in his mind and which he used to try to get rid of by exercising his mind

From Tesla's autobiography:

>To free myself of these tormenting appearances, I tried to concentrate my mind
 on something else I had seen...

Once he got control, he used visualization to help him with his inventions.

From Tesla's autobiography:

>When I get an idea, I start at once building it up in my imagination. I change
 the construction, make improvements and operate the device in my mind. It is
 absolutely immaterial to me whether I run my turbine in thought or test it in
 my shop. I even note if it is out of balance. There is no difference whatever;
 the results are the same. In this way I am able to rapidly develop and perfect
 a conception without touching anything. When I have gone so far as to embody in
 the invention every possible improvement I can think of and see no fault
 anywhere, I put into concrete form this final product of my brain.

Read more about it in his autobiography:
https://archive.org/details/NikolaTeslaTheStrangeLifeOfNikolaTesla/page/n9/mode/2up

**Comment by kaveh808** [comment]
I think kons-9 would be a good platform for mathematical visualizations, allowing for rapid and flexible development of such code in CL.

**Reply by theottm** [reply]
I totally agree. Lisp allows for interactive direct feedback. Also the system is extendable, allowing for special needs. Once the system can handle bigger size scenes, it can legit be used for everyday plotting as well as more complex visualizations.

**Reply by theottm** [reply]
I'll get to that parametrized surface feature once I done with exams :) that open doors.
